### PR TITLE
(ADT-DP) Add diagnostic scopes to all DigialTwins APIs

### PR DESCRIPTION
**Pull request by azabbasi**

### CODE REVIEW

**Review by timtay-microsoft — APPROVED** [review]

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs [hunk 1]
```diff
@@ -152,13 +152,25 @@ protected DigitalTwinsClient()
...
+
+                // Deserialize the stream into the specified type
+                T deserializedDigitalTwin = (T)await _objectSerializer.DeserializeAsync(digitalTwinStream, typeof(T), cancellationToken).ConfigureAwait(false);
 
-            return Response.FromValue<T>(deserializedDigitalTwin, digitalTwinStream.GetRawResponse());
+                return Response.FromValue<T>(deserializedDigitalTwin, digitalTwinStream.GetRawResponse());
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
```
Comment (abhipsaMisra) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:171]: is there a `scope.Stop()` for the happy path?

Comment (azabbasi) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:171]: There is no Stop API, but during disposal, the activity is implicitly stopped.

**Review by barustum — APPROVED** [review]

**Review by abhipsaMisra — APPROVED** [review]
One question - given that we are creating these scope instances every time the API is called, what would be the performance (memory) impact if I call an API in a loop 100 times? (for disposing - it'll wait until the gc collects it, which isn't always immediate, right?)

**Comment by azabbasi** [comment]
> One question - given that we are creating these scope instances every time the API is called, what would be the performance (memory) impact if I call an API in a loop 100 times? (for disposing - it'll wait until the gc collects it, which isn't always immediate, right?)

well, it won't wait for the gc since as soon as the method call goes out of scope the objects are disposed.

**Pull request merged**